src/open_clip/loss.py

```diff
@@ -119,3 +119,48 @@ def forward(self, image_features, text_features, logit_scale):
             F.cross_entropy(logits_per_text, labels)
             ) / 2
         return total_loss
+
+
+class CoCaLoss(ClipLoss):
+    def __init__(
+        self,
+        caption_loss_weight,
+        clip_loss_weight,
+        pad_id=-100,
+        local_loss=False,
+        gather_with_grad=False,
+        cache_labels=False,
+        rank=0,
+        world_size=1,
+        use_horovod=False,
+    ):
+        super().__init__(
+            local_loss=local_loss,
+            gather_with_grad=gather_with_grad,
+            cache_labels=cache_labels,
+            rank=rank,
+            world_size=world_size,
+            use_horovod=use_horovod
+        )
+
+        self.clip_loss_weight = clip_loss_weight
+        self.caption_loss_weight = caption_loss_weight
+        self.caption_loss = nn.CrossEntropyLoss(ignore_index=pad_id)
+
+    def forward(self, image_features, text_features, logits, labels, logit_scale):
+        clip_loss = super().forward(image_features, text_features, logit_scale)
+        clip_loss = self.clip_loss_weight * clip_loss
+
+        if self.world_size > 1:
```

Comment:
does the whole local_loss thing not apply here? why is this not like the corresponding part of the ClipLoss

Reply:
I tried, I hope I got it right, I don't know too much about this

Reply:
@iejMac I am replying to all comments I think need a reply and I tag you in the things that I think are more relevant / I understand less.

After thinking a bit more, I think `local_loss` is not really needed in the captioning loss, since, if I gather `logits`  or `labels` then either I gather also the other one or I could have avoided gathering in the first place as the loss is "one by one", while for clip_loss I believe it is implemented to have a smaller matrix product in case of large batch size?